src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -413,65 +312,167 @@ public ValueTask<(Solution solution, bool updated)> TryUpdateWorkspaceCurrentSol
+                        refCountedLazySolution.AddReference_WhileAlreadyHoldingLock();
+                        return refCountedLazySolution;
+                    }
+
+                    return null;
+                }
             }
 
-            public Task<Solution> Task => _task;
+            public async ValueTask<RefCountedSolution> SlowGetOrCreateSolutionAsync(
+                Checksum solutionChecksum, Func<CancellationToken, Task<Solution>> getSolutionAsync, CancellationToken cancellationToken)
+            {
+                using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    if (_solutionChecksumToLazySolution.TryGetValue(solutionChecksum, out var refCountedLazySolution))
+                    {
+                        refCountedLazySolution.AddReference_WhileAlreadyHoldingLock();
+                        return refCountedLazySolution;
+                    }
+
+                    // We're the first call that is asking about this checksum.  Create a lazy to compute it with a
+                    // refcount of 1 (for 'us').
+                    refCountedLazySolution = new RefCountedSolution(this, solutionChecksum, getSolutionAsync);
+                    Contract.ThrowIfFalse(refCountedLazySolution.RefCount == 1);
+                    _solutionChecksumToLazySolution.Add(solutionChecksum, refCountedLazySolution);
+                    return refCountedLazySolution;
+                }
+            }
 
-            void IDisposable.Dispose()
+            public async Task SetLastRequestedSolutionAsync(Checksum solutionChecksum, RefCountedSolution solution, CancellationToken cancellationToken)
             {
-                // This type is only used as IAsyncDisposable, but needs to implement IDisposable for wrapping within
-                // ReferenceCountedDisposable<T>.
-                throw ExceptionUtilities.Unreachable;
+                using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    var solutionToRelease = _lastRequestedSolution;
+
+                    solution.AddReference_WhileAlreadyHoldingLock();
+                    _lastRequestedSolution = solution;
+                    _lastRequestedChecksum = solutionChecksum;
+
+                    solutionToRelease?.Release_WhileAlreadyHoldingLock();
+                }
             }
 
-            public async ValueTask DisposeAsync()
+            public sealed class RefCountedSolution : IDisposable
             {
-                _cancellationTokenSource.Cancel();
-                _cancellationTokenSource.Dispose();
+                private readonly ChecksumToSolutionCache _cache;
+                private readonly Checksum _solutionChecksum;
+                private readonly CancellationTokenSource _cancellationTokenSource = new();
+                private readonly Task<Solution> _task;
+
+                private int _refCount = 1;
```

Comment:
moved away from ReferenceCountedDisposable.  It was adding a lot of complexity into the localized logic here.  It also didn't have semantics that we wanted.  Namely, RCD has the behavior that *any* release of hte ref count on a particular RCD will clear out that RCD (but the underlying object still only gets disposed when the refcount goes to 0).  In this domain we want to point at the ref-counted-entity in a few places (e.g. it can be in several cache dictionaries or direct pointers).  This meant it was trivial to end up with the RCD in a location where we have a refcount>0 but we still couldn't get to the underlying entity.